src/shaders/_prelude_fog.fragment.glsl

```diff
@@ -15,45 +25,52 @@ float fog_sky_blending(vec3 camera_dir) {
     return u_fog_opacity * exp(-3.0 * t * t);
 }
 
-float fog_opacity(vec3 pos) {
-    float depth = length(pos);
-    float start = u_fog_range.x;
-    float end = u_fog_range.y;
-
-    // The fog is not physically accurate, so we seek an expression which satisfies a
-    // couple basic constraints:
-    //   - opacity should be 0 at the near limit
-    //   - opacity should be 1 at the far limit
-    //   - the onset should have smooth derivatives to avoid a sharp band
-    // To this end, we use an (1 - e^x)^n, where n is set to 3 to ensure the
-    // function is C2 continuous at the onset. The fog is about 99% opaque at
-    // the far limit, so we simply scale it and clip to achieve 100% opacity.
-    // https://www.desmos.com/calculator/3taufutxid
-    const float decay = 5.5;
-    float falloff = max(0.0, 1.0 - exp(-decay * (depth - start) / (end - start)));
+// This function gives the fog opacity when strength is 1. Otherwise it's multiplied
+// by a smoothstep to a power to decrease the amount of fog relative to haze.
+// This function much match src/style/fog.js
+// See: https://www.desmos.com/calculator/3taufutxid
+float fog_opacity(float t) {
+    const float decay = 6.0;
+    float falloff = 1.0 - min(1.0, exp(-decay * t));
 
     // Cube without pow()
     falloff *= falloff * falloff;
 
     // Scale and clip to 1 at the far limit
-    falloff = min(1.0, 1.00747 * falloff);
+    return u_fog_opacity * min(1.0, 1.00747 * falloff);
+}
+
+float fog_opacity (vec3 pos) {
+    return fog_opacity((length(pos) - u_fog_range.x) / (u_fog_range.y - u_fog_range.x));
+}
+
+vec3 fog_apply(vec3 color, vec3 pos) {
+    // Map [near, far] to [0, 1]
+    float t = (length(pos) - u_fog_range.x) / (u_fog_range.y - u_fog_range.x);
```

Comment:
Should we disallow `u_fog_range.y == u_fog_range.x` from the style spec inputs? A min difference of 1 can be enough to prevent UB / division by 0 here.

It can be added in the validation errors in https://github.com/mapbox/mapbox-gl-js/blob/73c369fe5be7a88ebb6a9d90c602e2f7596c4db2/src/style-spec/validate/validate_fog.js#L21-L27

as in commit https://github.com/mapbox/mapbox-gl-js/commit/73c369fe5be7a88ebb6a9d90c602e2f7596c4db2.

Reply:
What about just changing greater than to greater than or equal to?

Reply:
Yes that works.